modules/table/src/main/java/org/apache/ignite/internal/table/distributed/storage/InternalTableImpl.java

```diff
@@ -302,4 +336,146 @@ private int partId(BinaryRow row) {
+         * Partition Scan Subscription.
+         */
+        private class PartitionScanSubscription implements Subscription {
+            /** */
+            private final Subscriber<? super BinaryRow> subscriber;
+
+            /** */
+            private final AtomicBoolean isCanceled;
+
+            /** Scan id to uniquely identify it on server side. */
+            private final IgniteUuid scanId;
+
+            /** Scan initial operation that created server cursor. */
+            private final CompletableFuture<Void> scanInitOp;
+
+            /**
+             * The constructor.
+             * @param subscriber The subscriber.
+             */
+            private PartitionScanSubscription(Subscriber<? super BinaryRow> subscriber) {
+                this.subscriber = subscriber;
+                this.isCanceled = new AtomicBoolean(false);
+                this.scanId = UUID_GENERATOR.randomUuid();
+                // TODO: IGNITE-15544 Close partition scans on node left.
+                this.scanInitOp = raftGrpSvc.run(new ScanInitCommand("", scanId));
+            }
+
+            /** {@inheritDoc} */
+            @Override public void request(long n) {
+                if (n <= 0) {
+                    cancel();
+
+                    subscriber.onError(new IllegalArgumentException(LoggerMessageHelper.
+                        format("Invalid requested amount of items [requested={}, minValue=1]", n))
+                    );
+                }
+
+                if (isCanceled.get())
+                    return;
+
+                final int internalBatchSize = Integer.MAX_VALUE;
+
+                for (int intBatchCnr = 0; intBatchCnr < (n / internalBatchSize); intBatchCnr++)
+                    scanBatch(internalBatchSize);
+
+                scanBatch((int)(n % internalBatchSize));
+            }
+
+            /** {@inheritDoc} */
+            @Override public void cancel() {
+                cancel(true);
+            }
+
+            /**
+             * Cancels given subscription and closes cursor if necessary.
+             *
+             * @param closeCursor If {@code true} closes inner storage scan.
+             */
+            private void cancel(boolean closeCursor) {
+                isCanceled.set(true);
+
+                if (closeCursor) {
+                    raftGrpSvc.run(new ScanCloseCommand(scanId)).exceptionally(closeT -> {
+                        LOG.warn("Unable to close scan.", closeT);
+
+                        return null;
+                    });
+                }
+            }
+
+            /**
+             * Requests and processes n requested elements where n is an integer.
+             *
+             * @param n Requested amount of items.
+             */
+            private void scanBatch(int n) {
+                if (isCanceled.get())
+                    return;
+
+                scanInitOp.thenCompose((none) -> raftGrpSvc.<MultiRowsResponse>run(new ScanRetrieveBatchCommand(n, scanId)))
```

Comment:
A race is possible here. The scan initialization could take some time, and caller could cancel the scan, thus lead as to a follow sequence of events: `SCAN_INIT`, `SCAN_CLOSE`, `SCAN_RETRIEVE`. We need to check `isCanceled` flag right before invocation of `raftGrpSvc .run(ScanRetrieveBatchCommand)`

Reply:
> We need to check isCanceled flag right before invocation of raftGrpSvc .run(ScanRetrieveBatchCommand)
Without extra synchronization that won't help.